version-check.js

```diff
@@ -0,0 +1,119 @@
+const https = require('https');
+const exec = require('child_process').exec;
+
+// TODO: switch to develop branch
+const mappingsSrc = 'https://raw.githubusercontent.com/service-mocker/service-mocker/ts-compatibility-check/ts-version-map.json';
+
+const MOCKER_MATCHED = 1 << 0;
+const TS_MATCHED = 1 << 1;
+
+function installDeps() {
+  const deps = Array.prototype.slice.call(arguments);
+  const missed = [];
+
+  deps.forEach((pkg) => {
+    try {
+      require.resolve(pkg);
+    } catch (e) {
+      missed.push(pkg);
+    }
+  });
+
+  if (!missed.length) {
+    return Promise.resolve();
+  }
+
+  return new Promise((resolve, reject) => {
+    exec(`npm install ${missed.join(' ')}`, (err, stdout, stderr) => {
+      if (err || stderr) {
+        reject(err || stderr);
+      } else {
+        resolve(stdout);
+      }
+    });
+  });
+}
+
+function fetchRemote() {
+  return new Promise((resolve, reject) => {
+    https.get(mappingsSrc, (res) => {
+      let rawData = '';
+
+      res.on('data', (chunk) => {
+        rawData += chunk;
+      });
+
+      res.on('end', () => {
+        try {
+          resolve(JSON.parse(rawData));
+        } catch (e) {
+          reject(e);
+        }
+      });
+    }).on('error', reject);
+  });
+}
+
+function compareVer(versions) {
+  const mockerVersion = require('./package.json').version;
+  const tsVersion = require('typescript').version;
+
+  return fetchRemote()
+    .then((data) => {
+      const semver = require('semver');
+
+      console.log('Compatibility List:', JSON.stringify(data.mappings, null, 2));
+
+      let finalState = 0;
+      let error = null;
+
+      data.mappings.some((mapping) => {
+        let state = 0;
+
+        if (semver.satisfies(mockerVersion, mapping['service-mocker'])) {
+          state |= MOCKER_MATCHED;
+          finalState |= MOCKER_MATCHED;
+        }
+
+        if (semver.satisfies(tsVersion, mapping['typescript'])) {
+          state |= TS_MATCHED;
+          finalState |= TS_MATCHED;
+        }
+
+        switch (state) {
+          case MOCKER_MATCHED & TS_MATCHED:
+            return true;
+
+          case MOCKER_MATCHED:
+            return false; // next
+
+          case TS_MATCHED:
+            error = new Error(`service-mocker@${mockerVersion} is not compatible with typescript@${tsVersion}, please use service-mocker@${mapping['service-mocker']}`);
+            return true;
+        }
+      });
+
+      if ((finalState & TS_MATCHED) === 0) {
+        error = new Error(`service-mocker dose not support typescript@${tsVersion} yet, please file an issue on https://github.com/service-mocker/service-mocker/issues.`);
+      }
+
+      if (error !== null) {
+        throw error;
+      }
+    });
+}
+
+try {
+  // check ts existence
+  require.resolve('typescript');
```

Comment:
Yeah, the case:

> run `npm install` in this repo.

Since [**npm3 does not install dependencies in a deterministic way**](https://docs.npmjs.com/how-npm-works/npm3-nondet#do-different-dependency-tree-structures-affect-my-app), if the installing order is:
```
install service-mocker  => postinstall of service-mocker => install typescript
```
`require.resolve('typescript')` will always throw an module not found error.

But `postinstall` should be executed after all packages installed:
https://github.com/npm/npm/blob/v4.5.0/lib/install.js#L292-L294

So I think it is ok here...(Never mind)